src/App.js

```diff
@@ -44,8 +90,164 @@ class App extends Component {
+
+        this.setState({
+            modalCategoryActive: true,
+            modalCategoryState: 'add',
+            categoryToUpdate: parent
+        });
+    }
+
+    handleUpdateCategory(node) {
+
+        this.setState({
+            modalCategoryActive: true,
+            modalCategoryState: 'update',
+            modalCategoryInputText: node.name,
+            categoryToUpdate: node
+        });
+    }
+
+    handleCancelModalCategory() {
+        this.setState({
+            modalCategoryActive: false,
+            modalCategoryState: null,
+            modalCategoryInputText: '',
+            categoryToUpdate: null
+        });
+    }
+
+    handleSubmitModalCategory(categoryName) {
+
+        if(this.state.modalCategoryState === 'update') {
+            const list1 = fromJS(this.state.categories);
+            const categoryIndex = this.state.categories.indexOf(this.state.categoryToUpdate);
+            const list2 = list1.updateIn([categoryIndex, 'name'], name => categoryName).toJS();
+
+            this.setState({
+                modalCategoryActive: false,
+                modalCategoryState: null,
+                categoryToUpdate: null,
+                categories: list2
+            });
+
+        } else if(this.state.modalCategoryState === 'add') {
+
+            if(categoryName === '') return;
+
+            const newCategory = {
+                id: this.state.nextCatId,
+                name: categoryName,
+                parentId: this.state.categoryToUpdate.id
+            };
+
+            let list2 = this.state.categories;
+            list2.unshift(newCategory);
+
+            this.setState({
+                modalCategoryActive: false,
+                modalCategoryState: null,
+                categoryToUpdate: null,
+                nextCatId: this.state.nextCatId + 1
+            });
+
+        }
+    }
+
+    handleInputChange() {
+
+        this.setState({
+            addCategoryInputText: this.categoriesComponent.addInput.value,
+            taskInputText: this.tasksComponent.taskInput.value,
+            searchInputText: this.searchInput.value,
+        });
+
+        if(this.state.modalCategoryActive) {
+            this.setState({
+                modalCategoryInputText: this.modalComponent.modalInput.value,
+            });
+        }
+    }
+
+    handleCountCategories() {
+        this.setState({
+            numOfCategories: this.state.numOfCategories + 1
+        });
+    }
+
+    handleShowDone() {
+        this.setState({ showDone: this.doneCheckbox.checked });
+    }
+
+    handleSetDone(task) {
+
+        let tasks = fromJS(this.state.tasks);
+        const taskIndex = this.state.tasks.indexOf(task);
+        const newTasks = tasks.updateIn([taskIndex, 'isDone'], isDone => !task.isDone).toJS();
+
+        this.setState({tasks: newTasks});
+    }
+
+    handleSaveTask(task) {
+
+        const tasks = List(this.state.tasks);
+        let taskIndex;
+        this.state.tasks.filter((item, index) => {
```

Comment:
There is a method for finding index - `findIndex`. See [MDN docs for details](https://developer.mozilla.org/pl/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex), but have in mind that not all browsers support it.

Anyway, since you are using Immutable List, you can use it's [findIndex method](https://facebook.github.io/immutable-js/docs/#/List/findIndex).

```js
const taskIndex = tasks.findIndex((item) => item.id === task.id);
```